Implement MapPhysicalMemory/UnmapPhysicalMemory

SciresM opened the pull request:
This implements svcMapPhysicalMemory/svcUnmapPhysicalMemory for Yuzu, which can be used to map memory at a desired address by games since 3.0.0.

It also properly parses SystemResourceSize from NPDM, and makes information available via svcGetInfo.

This is needed for games like Super Smash Bros. and Diablo 3 -- this PR's implementation does not run into the "ASCII reads" issue mentioned in the comments of #2626, which was caused by the following bugs in Yuzu's memory management that this PR also addresses:
* Yuzu's memory coalescing does not properly merge blocks. This results in a polluted address space/svcQueryMemory results that would be impossible to replicate on hardware, which can lead to game code making the wrong assumptions about memory layout.
  * This implements better merging for AllocatedMemoryBlocks.
* Yuzu's implementation of svcMirrorMemory unprotected the entire virtual memory range containing the range being mirrored. This could lead to games attempting to map data at that unprotected range/attempting to access that range after yuzu improperly unmapped it.
  * This PR fixes it by simply calling ReprotectRange instead of Reprotect.

lioncash commented on src/core/hle/kernel/process.h:
```diff
@@ -298,12 +299,16 @@ class Process final : public WaitObject {
+    /// Specifies additional memory to be reserved for the process's memory management by the
+    /// system. When this is non-zero, secure memory is allocated and used for page table allocation
+    /// instead of using the normal global page tables/memory block management.
+    u32 system_resource_size = 0;
```
This value should also be subtracted off the returned value in `GetTotalPhysicalMemoryAvailableWithoutMmHeap()` and `GetTotalPhysicalMemoryUsedWithoutMmHeap()`

lioncash commented on src/core/hle/kernel/vm_manager.cpp:
```diff
@@ -308,6 +309,258 @@ ResultVal<VAddr> VMManager::SetHeapSize(u64 size) {
+    }
+
+    // Check that we can map the memory we want.
+    const auto res_limit = Core::CurrentProcess()->GetResourceLimit();
```
Please do not use global accessors here.  The VMManager instance already has a reference to the system object as one of its members. Use that instead.

lioncash commented on src/core/hle/kernel/svc.cpp:
```diff
@@ -946,6 +960,86 @@ static ResultCode GetInfo(Core::System& system, u64* result, u64 info_id, u64 ha
+        return ERR_INVALID_MEMORY_RANGE;
+    }
+
+    auto* const current_process = Core::CurrentProcess();
```
This should be using the `system` parameter to retrieve the current process instead of using a global accessor function.

lioncash commented on src/core/hle/kernel/svc.cpp:
```diff
@@ -946,6 +960,86 @@ static ResultCode GetInfo(Core::System& system, u64* result, u64 info_id, u64 ha
+        return ERR_INVALID_MEMORY_RANGE;
+    }
+
+    auto* const current_process = Core::CurrentProcess();
```
Ditto

lioncash commented:
The commit message should also contain the extended description in the PR, so it's stored in the repository's history directly. Especially given it resolves a significant issue.

lioncash commented on src/core/hle/kernel/vm_manager.h:
```diff
@@ -450,6 +451,34 @@ class VMManager final {
+    ///
+    /// @note The destination address must lie within the Map region.
+    ///
+    /// @note This function requires SystemResourceSize is non-zero,
```
```suggestion
    /// @note This function requires that SystemResourceSize be non-zero,
```

Ditto below for UnmapPhysicalMemory too.

SciresM commented:
Force-pushed a commit message fix, and a commit that addresses review commentary other than 

```
These functions seem quite... cluttered and can be a lot to keep track of. I have a commit in a branch here: lioncash@61e1c6a that tries to factor out some of it. Maybe that could be used as a visual guide or something to make this slightly nicer to grok.
```

SciresM commented:
@lioncash I added the helpers from your branch -- I agree they're definitely cleaner.

I also removed the separate mapping + reprotecting logic in MapPhysicalMemory -- the Horizon Kernel does them separately to simplify unmapping on error, but there's no actual need for yuzu to do so and it simplifies logic to map as read-write instead of as None and reprotecting.

FernandoS27 commented on src/core/hle/kernel/vm_manager.cpp:
```diff
@@ -604,6 +765,57 @@ VMManager::VMAIter VMManager::MergeAdjacent(VMAIter iter) {
+            new_memory->insert(new_memory->end(), right.backing_block->begin() + right.offset,
+                               right.backing_block->begin() + right.offset + right.size);
+            left.backing_block = new_memory;
+            left.offset = 0;
```
When can this case happen? In the context of mapPhysicalMemory and others?

SciresM replied:
Memory block is mapped, and then part of the memory block is unmapped, or reprotected, or has SetMemoryAttribute called on it.

lioncash commented on src/video_core/memory_manager.cpp:
```diff
@@ -49,6 +52,12 @@ GPUVAddr MemoryManager::MapBufferEx(VAddr cpu_addr, u64 size) {
     const GPUVAddr gpu_addr{FindFreeRegion(address_space_base, aligned_size)};
 
     MapBackingMemory(gpu_addr, Memory::GetPointer(cpu_addr), aligned_size, cpu_addr);
+    ASSERT(Core::System::GetInstance()
```
Again, please don't use the global system accessor. If you need access to the system instance, pass down the reference via the constructor and then make use of it here. The GPU constructor where the memory manager instance is allocated already provides such a reference.

SciresM replied:
Okay.

lioncash replied:
Unresolved, as, while the reference itself was added to the class, it's not actually used in place of the global accessor function.

SciresM replied:
fixed + force-pushed. Sorry about that, hahaha.

FernandoS27 commented on src/core/hle/kernel/vm_manager.cpp:
```diff
@@ -604,6 +770,57 @@ VMManager::VMAIter VMManager::MergeAdjacent(VMAIter iter) {
+        UpdatePageTableForVMA(left);
+
+        // Update mappings for unicorn.
+        system.ArmInterface(0).UnmapMemory(left.base, left.size);
```
All these should be removed since unicorn mapping is no longer in master.

SciresM replied:
Done.

FernandoS27 commented:
@ogniK5377 @DarkLordZach request for review.
